mlflow/utils/databricks_utils.py

```diff
@@ -731,3 +727,68 @@ def get_databricks_env_vars(tracking_uri):
 
     return env_vars
+
+
+def _init_databricks_cli_config_provider(entry_point):
+    """
+    set a custom DatabricksConfigProvider with the hostname and token of the
+    user running the current command (achieved by looking at
+    PythonAccessibleThreadLocals.commandContext, via the already-exposed
+    NotebookUtils.getContext API)
+    """
+    notebook_utils = entry_point.getDbutils().notebook()
+
+    class DynamicConfigProvider(DatabricksConfigProvider):
+        def get_config(self):
+            logger = entry_point.getLogger()
+            try:
+                from dbruntime.databricks_repl_context import get_context
+
+                ctx = get_context()
+                if ctx and ctx.apiUrl and ctx.apiToken:
+                    return DatabricksConfig.from_token(
+                        host=ctx.apiUrl, token=ctx.apiToken, insecure=ctx.sslTrustAll
+                    )
+            except Exception as e:
+                print(  # noqa
+                    "Unexpected internal error while constructing `DatabricksConfig` "
+                    f"from REPL context: {e}",
+                    file=stderr,
+                )
+            # Invoking getContext() will attempt to find the credentials related to the
+            # current command execution, so it's critical that we execute it on every get_config().
+            api_url_option = notebook_utils.getContext().apiUrl()
+            api_url = api_url_option.get() if api_url_option.isDefined() else None
+            # Invoking getNonUcApiToken() will attempt to find the current credentials related to
+            # the current command execution and refresh it if its expired automatically,
+            # so it's critical that we execute it on every get_config().
+            api_token = None
+            try:
+                api_token = entry_point.getNonUcApiToken()
+            except Exception:
+                # Using apiToken from command context would return back the token which is not
+                # refreshed.
+                fallback_api_token_option = notebook_utils.getContext().apiToken()
+                logger.logUsage(
+                    "refreshableTokenNotFound",
+                    {"api_url": api_url},
+                    None,
+                )
+                if fallback_api_token_option.isDefined():
+                    api_token = fallback_api_token_option.get()
+
+            ssl_trust_all = entry_point.getDriverConf().workflowSslTrustAll()
+
+            # Fallback to the default behavior if we were unable to find a token.
+            if api_token is None or api_url is None:
+                return DefaultConfigProvider().get_config()
+
+            return DatabricksConfig.from_token(
+                host=api_url, token=api_token, insecure=ssl_trust_all
+            )
+
+    set_config_provider(DynamicConfigProvider())
+
+
+if is_in_databricks_runtime():
+    _init_databricks_cli_config_provider(_get_dbutils().entry_point)
```

Comment:
how long does this function take to finish?

Reply:
Do we need to run this here? Can we run this during REPL startup?

Reply:
very quick, less than 100ms

Reply:
note it is only executed in databricks runtime, it doesn't affect non-databricks environment

Reply:
Can we run this during REPL startup or not

Reply:
if we want to make it run during REPL startup, we have to add it in https://github.com/databricks/universe/blob/04ff2fffffd3d4fc27198e95f997aec8c5d75af0/spark/driver/python/python_shell/scripts/db_ipykernel_launcher.py#L118

but the function is only used by MLflow, so I prefer to move it and its invocation to mlflow,

and in MLlfow `_init_databricks_cli_config_provider` will only be invoked once when the module is loaded, so there is no perf difference compared with "run during REPL startup"